### PR TITLE
feat(EquipmentExporterPlugin): Add equipment exporter for Void simulator

### PR DESCRIPTION
**Pull request by AmVoidGuy**
Feel free to change anything as I know it is not the cleanest implementation (it's functional at least), but this is just a new plugin for the combat sim which will export all of your current equipment and levels. 

Notes:
- Still needs logic for both consumable and ability changed messages so that they are automatically updated without needing to reload the webpage.
- Doesn't export custom triggers (but the combat sim doesn't currently import those anyways)
- Just exports the default fly fight and 100 hours simulation time which I think is fine for this purpose but maybe can be improved upon in the future.

### CODE REVIEW

File: package-lock.json
Comment (Ishadijcks) [inline package-lock.json:1]: The `package-lock` should not have changed. Did you do anything more than an `npm install`?

File: src/MooLite/core/inventory/items/EquippedItems.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class EquippedItems {
+    name;
+    level;
+    constructor(name, level) {
+        this.name = name;
+        this.level = level;
+    }
+}
+export default EquippedItems;
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/items/EquippedItems.ts:9]: Why do we need a specific interface for this? Is this not just a `CharacterItem`

File: src/MooLite/core/inventory/items/CombatConsumable.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { ItemHrid } from "src/MooLite/core/inventory/ItemHrid";
+
+export interface CombatConsumable {
+    itemHrid: ItemHrid;
+}
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/items/CombatConsumable.ts:5]: Why do we need a specific interface for this? Is this not just a CharacterItem

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import { MooLitePlugin } from "src/MooLite/core/plugins/MooLitePlugin";
+import { MooLiteTab } from "src/MooLite/core/plugins/MooLiteTab";
+import EquipmentExporterPluginDisplay from "src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue";
+import EquippedItems from "src/MooLite/core/inventory/items/EquippedItems";
+import {CharacterAbility} from "src/MooLite/core/abilities/CharacterAbility";
+
+export class EquipmentExporterPlugin extends MooLitePlugin {
+    name: string = "Equipment Exporter";
+    key = "equipment-exporter";
+    description: string = "Export your equipment details for the Combat Sim";
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts:11]: Maybe add some clarification on that it is _your_ combat sim :)

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        componentName: "EquipmentExporterPluginDisplay",
+        component: markRaw(EquipmentExporterPluginDisplay),
+    };
+
+    getEquipmentSummary(): Array<EquippedItems> | null {
+        const equippedItems = new Array<EquippedItems>();
+        const abilities = this._game.combat.getCombatUnit().combatAbilities;
+        const combatFood = this._game.combat.getCombatUnit().combatConsumables;
+        const allItems = this._game.inventory._characterItems;
+        ["head", "body", "legs", "feet", "hands", "main_hand", "two_hand", "off_hand", "pouch", "neck", "earrings", "ring"].forEach((location) => {
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts:25]: You can grab these from the `ItemLocationTypeMap` so they stay in sync with game updates automatically

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            let equipmentLocation = "/item_locations/" + location;
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts:26]: Then this wouldn't be needed either

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+const exportDetails = () => {
+  try {
+    const equipmentData = props.plugin.getEquipmentSummary();
+    let abilityLevel0 = "1";
+    let abilityLevel1 = "1";
+    let abilityLevel2 = "1";
+    let abilityLevel3 = "1";
+    let drinks = [];
+    let food = [];
+    let abilities = ["", "", "", ""];
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue:23]: We can get the length of this array from the abilities level threshold

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  return props.plugin.getEquipmentSummary();
+});
+
+const exportDetails = () => {
+  try {
+    const equipmentData = props.plugin.getEquipmentSummary();
+    let abilityLevel0 = "1";
+    let abilityLevel1 = "1";
+    let abilityLevel2 = "1";
+    let abilityLevel3 = "1";
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue:20]: Array?

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+const props = defineProps<{
+  plugin: EquipmentExporterPlugin;
+}>();
+
+const equipmentData = computed(() => {
+  return props.plugin.getEquipmentSummary();
+});
+
+const exportDetails = () => {
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue:14]: This giant block of logic is better suited for the actual plugin, so we can keep the Vue component for presentation logic only.

Comment (Ishadijcks) [reply src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue:14]: If the plugin gets too big, you are free to create addition helper `.ts` files

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+      "/equipment_types/feet": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/hands": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/main_hand": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/two_hand": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/off_hand": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/pouch": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/neck": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/earrings": {hrid: "", enhancementLevel: 0},
+      "/equipment_types/ring": {hrid: "", enhancementLevel: 0},
+    };
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue:46]: Can also loop over EquipmentTypeHrid here

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        } else if(equipmentData[i].name.includes("_bulwark") || equipmentData[i].name.includes("_slasher")){
+          equipment["/equipment_types/main_hand"].hrid = equipmentData[i].name;
+          equipment["/equipment_types/main_hand"].enhancementLevel = equipmentData[i].level;
+        } else if(equipmentData[i].name.includes("_buckler") || equipmentData[i].name.includes("_dirk") || equipmentData[i].name.includes("_shield")) {
+          equipment["/equipment_types/off_hand"].hrid = equipmentData[i].name;
+          equipment["/equipment_types/off_hand"].enhancementLevel = equipmentData[i].level;
+        } else if(equipmentData[i].name.includes("coffee")) {
+          drinks.push(equipmentData[i].name);
+        } else if(equipmentData[i].name.includes("donut") || equipmentData[i].name.includes("cake") || equipmentData[i].name.includes("gummy") || equipmentData[i].name.includes("yogurt")) {
+          food.push(equipmentData[i].name);
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue:88]: This entire if/else monstrosity is unneeded. You can get the location where stuff is equipped from the `ItemDetail`!

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+      if(equipment[equip].hrid === ""){
+        equipment[equip] = null;
+      }
+    }
+    while(drinks.length < 3){
+      drinks.push("");
+    }
+    while(food.length < 3){
+      food.push("");
+    }
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPluginDisplay.vue:102]: I'd prefer not to use while loops for the chance of infinite loops. Also we can get the max length from game data

**Review by Ishadijcks — CHANGES_REQUESTED** [review]
Thanks for your pull request 😄 
Overall it looks good, and with some changes I'll happily merge it 🥳 

I've added a bunch of individual comments, but I'd like to highlight my general thoughts and reasoning here:

- Do not change the `package-lock`. It means that some of our actual dependencies got updated and could possibly download dangerous code. In general I never merge any code that updates `package-lock` unless I can tell exactly what it does (which I can't at 720 lines)

- Separate presentation and business logic. You perform some logic in the `.vue` component that is better suited for the actual plugin, or a helper class if we can better organise it that way.

- Don't hardcode what we can infer from game data. There are many instances where you loop over hardcoded item locations, or assume the length of certain arrays (max consumables). It is highly likely these things will change in a future game update, so it is best to be prepared. Currently there is nothing hardcoded so plugins won't break. If mo9 adds another skill, the xp tracker will still work without adding any code. I think many such improvements could be made in this pull request. If game data is missing and you need it for a feature, make an issue and I'll add it ASAP.

Please don't let this feedback discourage you, it is a great feature that will make the combat simulator much easier to use 😄 If you have any questions (or disagree!), please let me know, and we can work together on getting this merged!

File: package-lock.json
Comment (AmVoidGuy) [reply package-lock.json:1]: Should be reverted now. It just tacked on dev: true for all the packages when I was running it locally to develop and accidentally threw it in the commit

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts (hunk 4 above)
Comment (AmVoidGuy) [reply src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts]: I think you meant the ItemLocationDetailMap, which was commented out, I implemented it and it seems to be working fine.

**Comment by AmVoidGuy** [comment]
I'm not sure if the logic behind the ability slots updating works correctly if you have multiple abilities because I'm no combat chad with enough int to test. I think I covered all of your comments though. It definitely at least looks a lot better now

File: src/MooLite/core/abilities/Abilities.ts [hunk 12]
```diff
@@ -12,12 +12,17 @@ export interface AbilityLvlGained extends CharacterAbility {
     delta: number;
 }
 
+export interface AbilitySlotChanged extends CharacterAbility {
+    delta: number;
```
Comment (Ishadijcks) [inline src/MooLite/core/abilities/Abilities.ts:16]: Rename this to `indexChanged`

File: src/MooLite/core/abilities/Abilities.ts [hunk 13]
```diff
@@ -72,6 +82,12 @@ export class Abilities {
                     delta: lvlChanged,
                 });
             }
+            if (slotChanged !== 0) {
+                this._onAbilitySlotChanged.dispatch({
+                    ...(this._characterAbilities[ability.abilityHrid] as CharacterAbility),
```
Comment (Ishadijcks) [inline src/MooLite/core/abilities/Abilities.ts:87]: Is the case needed here?

File: src/MooLite/core/inventory/Inventory.ts [hunk 14]
```diff
@@ -6,6 +6,8 @@ import { ItemCategoryDetail } from "src/MooLite/core/inventory/items/ItemCategor
 import { SimpleEventDispatcher } from "strongly-typed-events";
 import { ActionTypeHrid } from "src/MooLite/core/actions/ActionTypeHrid";
 import { CharacterConsumable } from "src/MooLite/core/inventory/items/CharacterConsumable";
+import {ItemLocationHrid} from "src/MooLite/core/inventory/ItemLocationHrid";
+import {ItemLocationDetail} from "src/MooLite/core/inventory/ItemLocationDetail";
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/Inventory.ts:10]: Don't forget to run `npm run lint` :)

File: src/MooLite/core/inventory/Inventory.ts [hunk 15]
```diff
@@ -93,4 +98,12 @@ export class Inventory {
     public isAbilityBook(itemHrid: ItemHrid): boolean {
         return this.itemDetailMap[itemHrid].abilityBookDetail?.abilityHrid.toString().length > 0;
     }
+
+    public get_equippedFood(): CharacterConsumable[] {
+        return this._actionTypeFoodSlotsMap["/action_types/combat"];
+    }
+
+    public get_equippedDrinks(): CharacterConsumable[] {
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/Inventory.ts:106]: Please rename this and above to `getEquippedDrinks`

File: src/MooLite/core/inventory/Inventory.ts [hunk 16]
```diff
@@ -93,4 +98,12 @@ export class Inventory {
     public isAbilityBook(itemHrid: ItemHrid): boolean {
         return this.itemDetailMap[itemHrid].abilityBookDetail?.abilityHrid.toString().length > 0;
     }
+
+    public get_equippedFood(): CharacterConsumable[] {
+        return this._actionTypeFoodSlotsMap["/action_types/combat"];
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/Inventory.ts:103]: Hmm is there no way we can do this without hardcoding `"/action_types/combat"`?

File: src/MooLite/core/plugins/MooLitePlugin.ts [hunk 17]
```diff
@@ -96,6 +96,8 @@ export abstract class MooLitePlugin {
 
     onAbilityLvlGained?(gains: AbilityXpGained): void;
 
+    onAbilitySlotChanged?(gains: AbilitySlotChanged): void;
```
Comment (Ishadijcks) [inline src/MooLite/core/plugins/MooLitePlugin.ts:99]: Please rename the argument to `slotChanged`

File: src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+    tab: MooLiteTab = {
+        icon: "🗡",
+        pluginName: this.name,
+        componentName: "EquipmentExporterPluginDisplay",
+        component: markRaw(EquipmentExporterPluginDisplay),
+    };
+
+    getEquipmentSummary(): Array<CharacterItem> | null {
+        const equippedItems = new Array<CharacterItem>();
```
Comment (Ishadijcks) [inline src/MooLite/plugins/EquipmentExporter/EquipmentExporterPlugin.ts:22]: Please use `[]` for array instantiation and typing

**Review by Ishadijcks — CHANGES_REQUESTED** [review]
Much cleaner :)

Some small changes left but nothing major!

File: src/MooLite/core/abilities/Abilities.ts (hunk 13 above)
Comment (AmVoidGuy) [reply src/MooLite/core/abilities/Abilities.ts:87]: Nope whoops, turns out it isn't

**Comment by AmVoidGuy** [comment]
> I'm not sure if the logic behind the ability slots updating works correctly if you have multiple abilities because I'm no combat chad with enough int to test. I think I covered all of your comments though. It definitely at least looks a lot better now

Also was able to test this and it does work as expected.

File: src/MooLite/core/inventory/ItemLocationHrid.ts [hunk 19]
```diff
@@ -1,3 +1,4 @@
-export enum ItemLocationHrid {
-    Inventory = "/item_locations/inventory",
-}
+export enum ItemLocationHrid {}
+//There is already an /item_locations/inventory in the ItemLocationDetailMap, so not sure why we would need this.
+//Also since it is a String, it is messing with searching by index since the rest are Hrids.
+//Inventory = "/item_locations/inventory",
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/ItemLocationHrid.ts:4]:
```suggestion
```

File: src/MooLite/core/inventory/ItemLocationHrid.ts [hunk 20]
```diff
@@ -1,3 +1,4 @@
-export enum ItemLocationHrid {
-    Inventory = "/item_locations/inventory",
-}
+export enum ItemLocationHrid {}
+//There is already an /item_locations/inventory in the ItemLocationDetailMap, so not sure why we would need this.
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/ItemLocationHrid.ts:2]:
```suggestion
```

File: src/MooLite/core/inventory/ItemLocationHrid.ts [hunk 21]
```diff
@@ -1,3 +1,4 @@
...
+//Also since it is a String, it is messing with searching by index since the rest are Hrids.
```
Comment (Ishadijcks) [inline src/MooLite/core/inventory/ItemLocationHrid.ts:3]:
```suggestion
```

**Review by Ishadijcks — DISMISSED** [review]

**Pull request merged**